packages/grid/_modules_/grid/hooks/features/groupingColumns/useGridGroupingColumns.tsx

```diff
@@ -0,0 +1,480 @@
+    ...state,
+    groupingColumns: {
+      model: props.groupingColumnsModel ?? props.initialState?.groupingColumns?.model ?? [],
+    },
+  }));
+
+  const [, setGridState, forceUpdate] = useGridState(apiRef);
+
+  apiRef.current.unstable_updateControlState({
+    stateId: 'groupingColumns',
+    propModel: props.groupingColumnsModel,
+    propOnChange: props.onGroupingColumnsModelChange,
+    stateSelector: gridGroupingColumnsModelSelector,
+    changeEvent: GridEvents.groupingColumnsModelChange,
+  });
+
+  /**
+   * ROW GROUPING
+   */
+  const sanitizedGroupingColumnsModelOnLastRowPreProcessing =
+    React.useRef<GridGroupingColumnsModel>([]);
+
+  const updateRowGrouping = React.useCallback(() => {
+    const groupRows: GridRowGroupingPreProcessing = (params) => {
+      const groupingColumnsModel = gridGroupingColumnsSanitizedModelSelector(apiRef.current.state);
+      const columnsLookup = gridColumnLookupSelector(apiRef.current.state);
+      sanitizedGroupingColumnsModelOnLastRowPreProcessing.current = groupingColumnsModel;
+
+      if (props.disableGroupingColumns || groupingColumnsModel.length === 0) {
+        return null;
+      }
+
+      const distinctValues: { [field: string]: { map: { [val: string]: boolean }; list: any[] } } =
+        Object.fromEntries(
+          groupingColumnsModel.map((groupingField) => [groupingField, { map: {}, list: [] }]),
+        );
+
+      const getCellGroupingCriteria = ({
+        row,
+        id,
+        colDef,
+      }: {
+        row: GridRowModel;
+        id: GridRowId;
+        colDef: GridColDef;
+      }) => {
+        let value: GridCellValue;
+        if (colDef.valueGetter) {
+          const valueGetterParams: GridValueGetterSimpleParams = {
+            colDef,
+            field: colDef.field,
+            id,
+            row,
+            rowNode: {
+              isAutoGenerated: false,
+              id,
+            },
+          };
+          value = colDef.valueGetter(valueGetterParams);
+        } else {
+          value = row[colDef.field];
+        }
+
+        let key: GridKeyValue | null | undefined;
+        if (colDef.keyGetter) {
+          key = colDef.keyGetter({
+            value,
+            id,
+            field: colDef.field,
+          });
+        } else {
+          key = value as GridKeyValue;
+        }
+
+        return {
+          key,
+          field: colDef.field,
+        };
+      };
+
+      params.ids.forEach((rowId) => {
+        const row = params.idRowsLookup[rowId];
+
+        groupingColumnsModel.forEach((groupingCriteria) => {
+          const { key } = getCellGroupingCriteria({
+            row,
+            id: rowId,
+            colDef: columnsLookup[groupingCriteria],
+          });
+          const groupingFieldsDistinctKeys = distinctValues[groupingCriteria];
+
+          if (key != null && !groupingFieldsDistinctKeys.map[key.toString()]) {
+            groupingFieldsDistinctKeys.map[key.toString()] = true;
```

Comment:
We use a map to avoid the linear complexity of looking if the key is already registered.